apiserver/facades/controller/crossmodelsecrets/crossmodelsecrets.go

```diff
@@ -285,28 +278,49 @@ func (s *CrossModelSecretsAPI) getBackend(ctx stdcontext.Context, modelUUID stri
 }
 
 // canRead returns true if the specified entity can read the secret.
-func (s *CrossModelSecretsAPI) canRead(secretsConsumer SecretsConsumer, uri *coresecrets.URI, entity names.Tag) bool {
-	s.logger.Debugf("check %s can read secret %s", entity, uri.ID)
-	hasRole, _ := secretsConsumer.SecretAccess(uri, entity)
+func (s *CrossModelSecretsAPI) canRead(ctx stdcontext.Context, secretService SecretService, uri *coresecrets.URI, unit names.UnitTag) bool {
+	s.logger.Debugf("check %s can read secret %s", unit, uri.ID)
+	hasRole, _ := secretService.GetSecretAccess(ctx, uri, secretservice.SecretAccessor{
+		Kind: secretservice.UnitAccessor,
+		ID:   unit.Id(),
+	})
 	if hasRole.Allowed(coresecrets.RoleView) {
 		return true
 	}
 
-	// Unit access not granted, see if app access is granted.
-	if entity.Kind() != names.UnitTagKind {
-		return false
-	}
-	appName, _ := names.UnitApplication(entity.Id())
-	hasRole, _ = secretsConsumer.SecretAccess(uri, names.NewApplicationTag(appName))
+	appName, _ := names.UnitApplication(unit.Id())
+	hasRole, _ = secretService.GetSecretAccess(ctx, uri, secretservice.SecretAccessor{
+		Kind: secretservice.ApplicationAccessor,
+		ID:   appName,
+	})
 	return hasRole.Allowed(coresecrets.RoleView)
 }
 
-func (s *CrossModelSecretsAPI) accessScope(secretsConsumer SecretsConsumer, uri *coresecrets.URI, entity names.Tag) (names.Tag, error) {
-	s.logger.Debugf("scope for %q on secret %s", entity, uri.ID)
-	scope, err := secretsConsumer.SecretAccessScope(uri, entity)
-	if err == nil || !errors.Is(err, errors.NotFound) || entity.Kind() != names.UnitTagKind {
-		return scope, errors.Trace(err)
+func tagFromAccessScope(scope secretservice.SecretAccessScope) names.Tag {
+	switch scope.Kind {
+	case secretservice.ApplicationAccessScope:
+		return names.NewApplicationTag(scope.ID)
+	case secretservice.UnitAccessScope:
+		return names.NewUnitTag(scope.ID)
+	case secretservice.RelationAccessScope:
+		return names.NewRelationTag(scope.ID)
+	}
+	return nil
+}
+
+func (s *CrossModelSecretsAPI) accessScope(ctx stdcontext.Context, secretService SecretService, uri *coresecrets.URI, unit names.UnitTag) (names.Tag, error) {
+	s.logger.Debugf("scope for %q on secret %s", unit, uri.ID)
+	scope, err := secretService.GetSecretAccessScope(ctx, uri, secretservice.SecretAccessor{
+		Kind: secretservice.UnitAccessor,
+		ID:   unit.Id(),
+	})
+	if err == nil || !errors.Is(err, errors.NotFound) {
```

Comment:
Does the GetSecretAccessScope return a domain-level error?
And also should we translate the not found error to permission dinied error?

Reply:
This method is not implemented yet, so there's no new error available. When the method gets implemented, this will be fixed.